File: webapp/src/blocks/page.ts
// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

// Pages feature — TS model (skeleton).
// See docs/PAGES_PLAN.md for the data-model rationale.

export type Page = {
    id: string
    teamId: string
    parentId: string  // '' for top-level pages, else parent page id
    title: string
    icon: string
    cover: string
    sortOrder: number
    createdBy: string
    modifiedBy: string
    createAt: number
    updateAt: number
    deleteAt: number
}

export type BoardPageRef = {
    boardId: string
    pageId: string
    sortOrder: number
    label: string
    addedBy: string
    addedAt: number
}

export type PageBoardRef = {
    pageId: string
    boardId: string
    label: string
    addedBy: string
    addedAt: number
}

export type PagePatch = Partial<Pick<Page, 'title' | 'icon' | 'cover' | 'parentId' | 'sortOrder'>>

// Tiptap document JSON shape (kept loose; Tiptap exports a richer type).
export type TiptapDoc = {
    type: 'doc'
    content?: unknown[]
}

export type PageContent = {
    pageId: string
    tiptapJson?: TiptapDoc
    yjsState?: Uint8Array
    yjsUpdatesCount: number
    lastSnapshotAt: number
    updateAt: number
    updateBy: string
}

// Phase 2 — per-page ACL override.
export type PageMember = {
    pageId: string
    userId: string
    schemeAdmin: boolean
    schemeEditor: boolean
    schemeCommenter: boolean
    schemeViewer: boolean
}

export type PageChannelLink = {
    pageId: string
    channelId: string
    pinnedBy: string
    pinnedAt: number
}

export const createEmptyPage = (teamId: string, parentId?: string): Page => ({
    id: '',
    teamId,
    parentId: parentId || '',
    title: '',
    icon: '',
    cover: '',
    sortOrder: 0,
    createdBy: '',
    modifiedBy: '',
    createAt: 0,
    updateAt: 0,
    deleteAt: 0,
})

export const createEmptyTiptapDoc = (): TiptapDoc => ({
    type: 'doc',
    content: [{type: 'paragraph'}],
})

export const MAX_PAGE_DEPTH = 10
export const MAX_CHILDREN_PER_PARENT = 1000
export const MAX_SLUG_LENGTH = 50

File: server/model/page.go
// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package model

import "encoding/json"

// Page is a team-scoped collaborative document. Pages live in their own table
// (not under boards) and form a tree via parent_id (self-reference).
//
// Boards and pages cross-reference via board_page_refs and page_board_refs;
// neither owns the other.
//
// See docs/PAGES_PLAN.md (Model Y).
type Page struct {
	// ID is the page's UUID.
	ID string `json:"id"`

	// TeamID is the Mattermost team that owns this page.
	TeamID string `json:"teamId"`

	// ParentID is the parent page's ID. Empty string for top-level pages.
	ParentID string `json:"parentId"`

	// Title is shown in the sidebar tree and used to derive the URL slug.
	Title string `json:"title"`

	// Icon is an optional emoji or icon identifier shown next to the title.
	Icon string `json:"icon"`

	// Cover is an optional cover image URL.
	Cover string `json:"cover"`

	// SortOrder controls sibling order within a parent (smaller first).
	SortOrder int64 `json:"sortOrder"`

	// CreatedBy / ModifiedBy are user IDs.
	CreatedBy  string `json:"createdBy"`
	ModifiedBy string `json:"modifiedBy"`

	CreateAt int64 `json:"createAt"`
	UpdateAt int64 `json:"updateAt"`
	DeleteAt int64 `json:"deleteAt"`
}

// PagePatch describes a partial update to a Page.
type PagePatch struct {
	Title     *string `json:"title,omitempty"`
	Icon      *string `json:"icon,omitempty"`
	Cover     *string `json:"cover,omitempty"`
	ParentID  *string `json:"parentId,omitempty"`
	SortOrder *int64  `json:"sortOrder,omitempty"`
}

// PageContent holds the rich-text body of a Page.
//
// Phase 1: only TiptapJSON is used (single-editor mode).
// Phase 2: YjsState (compacted snapshot) and incremental updates in
// page_yjs_updates become the source of truth; TiptapJSON is rendered from
// the Yjs state on snapshot.
//
// TiptapJSON uses json.RawMessage so it serializes as raw JSON to clients.
// YjsState stays []byte — it's binary, base64 wire encoding is correct.
type PageContent struct {
	PageID          string          `json:"pageId"`
	TiptapJSON      json.RawMessage `json:"tiptapJson,omitempty"`
	YjsState        []byte          `json:"yjsState,omitempty"`
	YjsUpdatesCount int             `json:"yjsUpdatesCount"`
	LastSnapshotAt  int64           `json:"lastSnapshotAt"`
	UpdateAt        int64           `json:"updateAt"`
	UpdateBy        string          `json:"updateBy"`
}

// PageMember represents an explicit ACL entry for a page (Phase 2).
//
// Absence of a row means the page inherits from its nearest ancestor with a
// non-empty page_acl, falling back to the team-level permission baseline.
type PageMember struct {
	PageID          string `json:"pageId"`
	UserID          string `json:"userId"`
	SchemeAdmin     bool   `json:"schemeAdmin"`
	SchemeEditor    bool   `json:"schemeEditor"`
	SchemeCommenter bool   `json:"schemeCommenter"`
	SchemeViewer    bool   `json:"schemeViewer"`
}

// PageChannelLink ties a page to one or more Mattermost channels (many-to-many).
// Used to render "Pages in this channel" in the channel header.
type PageChannelLink struct {
	PageID    string `json:"pageId"`
	ChannelID string `json:"channelId"`
	PinnedBy  string `json:"pinnedBy"`
	PinnedAt  int64  `json:"pinnedAt"`
}

// BoardPageRef — a board references a page (board → page link).
type BoardPageRef struct {
	BoardID   string `json:"boardId"`
	PageID    string `json:"pageId"`
	SortOrder int64  `json:"sortOrder"`
	Label     string `json:"label"`
	AddedBy   string `json:"addedBy"`
	AddedAt   int64  `json:"addedAt"`
}

// PageBoardRef — a page references a board (page → board link).
type PageBoardRef struct {
	PageID  string `json:"pageId"`
	BoardID string `json:"boardId"`
	Label   string `json:"label"`
	AddedBy string `json:"addedBy"`
	AddedAt int64  `json:"addedAt"`
}

// QueryPageOptions filters page lookups.
type QueryPageOptions struct {
	TeamID         string // required
	ParentID       string // optional: list children of this parent ('' = top-level)
	IncludeDeleted bool
	Limit          uint64
}
